From 55a7a240c6eb42c1a1908fba5270a8ba646841d1 Mon Sep 17 00:00:00 2001
From: Athenas Jimenez Gonzalez <athenas.jimenez.gonzalez@intel.com>
Date: Wed, 2 May 2018 16:55:31 +0000
Subject: [PATCH] debug fma v2

---
 gcc/tree-ssa-math-opts.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/gcc/tree-ssa-math-opts.c b/gcc/tree-ssa-math-opts.c
index 16d9399af..4272c7bf3 100644
--- a/gcc/tree-ssa-math-opts.c
+++ b/gcc/tree-ssa-math-opts.c
@@ -2841,7 +2841,7 @@ last_fma_candidate_feeds_initial_phi (fma_deferring_state *state,
   and if we later discover an FMA candidate that is not part of such a chain,
   we go back and perform all deferred past candidates.  */
 
-static bool
+bool
 convert_mult_to_fma (gimple *mul_stmt, tree op1, tree op2,
 		     fma_deferring_state *state)
 {
-- 
2.17.0

