--- gcc-11.1.0/gcc/common.opt~	2021-04-27 10:00:13.000000000 +0000
+++ gcc-11.1.0/gcc/common.opt	2021-08-12 16:53:38.092485460 +0000
@@ -3175,7 +3175,7 @@
 Generate debug information in default version of DWARF format.
 
 gdwarf-
-Common Driver Joined UInteger Var(dwarf_version) Init(5) Negative(gstabs)
+Common Driver Joined UInteger Var(dwarf_version) Init(4) Negative(gstabs)
 Generate debug information in DWARF v2 (or later) format.
 
 gdwarf32
