--- gcc-5.3.0/gcc/fortran/options.c~	2015-02-04 17:38:48.000000000 -0500
+++ gcc-5.3.0/gcc/fortran/options.c	2016-03-09 12:00:49.444950867 -0500
@@ -79,6 +79,10 @@
   opts->frontend_set_flag_errno_math = true;
   opts->x_flag_associative_math = -1;
   opts->frontend_set_flag_associative_math = true;
+  opts->x_flag_tree_vectorize = 1;
+  opts->x_flag_tree_loop_vectorize = 1;
+  //opts->x_ix86_isa_flags |= OPTION_MASK_ISA_AVX;
+  //opts->x_ix86_isa_flags_explicit |= OPTION_MASK_ISA_AVX;
 }
 
 /* Get ready for options handling. Keep in sync with
