--- gcc-12.1.0/gcc/config/i386/gnu-user-common.h~	2022-05-06 07:30:56.000000000 +0000
+++ gcc-12.1.0/gcc/config/i386/gnu-user-common.h	2022-12-16 15:49:36.568801362 +0000
@@ -47,7 +47,7 @@
 
 /* Similar to standard GNU userspace, but adding -ffast-math support.  */
 #define GNU_USER_TARGET_MATHFILE_SPEC \
-  "%{Ofast|ffast-math|funsafe-math-optimizations:crtfastmath.o%s} \
+  "%{!mno-daz-ftz:crtfastmath.o%s} \
    %{mpc32:crtprec32.o%s} \
    %{mpc64:crtprec64.o%s} \
    %{mpc80:crtprec80.o%s}"
--- gcc-12.1.0/gcc/config/i386/i386.opt~	2022-05-06 07:30:57.000000000 +0000
+++ gcc-12.1.0/gcc/config/i386/i386.opt	2022-12-16 15:50:33.234422758 +0000
@@ -870,6 +870,10 @@
 Target Mask(ISA_CLWB) Var(ix86_isa_flags) Save
 Support CLWB instruction.
 
+mdaz-ftz
+Target
+Set the FTZ and DAZ Flags.
+
 mpcommit
 Target WarnRemoved
 
