rguenth (1):
      2019-05-03  Richard Biener  <rguenther@suse.de>

diff --git a/gcc/ChangeLog b/gcc/ChangeLog
index e9a3fa34129..fe1b698c195 100644
--- a/gcc/ChangeLog
+++ b/gcc/ChangeLog
@@ -1,3 +1,9 @@
+2019-05-03  Richard Biener  <rguenther@suse.de>
+
+	PR tree-optimization/90316
+	* tree-ssa-pre.c (pass_pre::execute): Re-compute DOM fast queries
+	before running VN.
+
 2019-05-03  Jakub Jelinek  <jakub@redhat.com>
 
 	* BASE-VER: Set to 9.1.1.
diff --git a/gcc/tree-ssa-pre.c b/gcc/tree-ssa-pre.c
index 7bb2cf1e220..e1c75f8e906 100644
--- a/gcc/tree-ssa-pre.c
+++ b/gcc/tree-ssa-pre.c
@@ -4197,6 +4197,7 @@ pass_pre::execute (function *fun)
   loop_optimizer_init (LOOPS_NORMAL);
   split_critical_edges ();
   scev_initialize ();
+  calculate_dominance_info (CDI_DOMINATORS);
 
   run_rpo_vn (VN_WALK);
 
