--- gcc-7.2.0/gcc/config/i386/avx512fintrin.h~	2017-04-10 06:53:28.000000000 +0000
+++ gcc-7.2.0/gcc/config/i386/avx512fintrin.h	2017-10-22 15:32:30.510498526 +0000
@@ -352,7 +352,7 @@
 __attribute__ ((__gnu_inline__, __always_inline__, __artificial__))
 _mm512_store_pd (void *__P, __m512d __A)
 {
-  *(__m512d *) __P = __A;
+  *(__m512d_u *) __P = __A;
 }
 
 extern __inline void
@@ -393,7 +393,7 @@
 __attribute__ ((__gnu_inline__, __always_inline__, __artificial__))
 _mm512_store_ps (void *__P, __m512 __A)
 {
-  *(__m512 *) __P = __A;
+  *(__m512_u *) __P = __A;
 }
 
 extern __inline void
@@ -453,7 +453,7 @@
 __attribute__ ((__gnu_inline__, __always_inline__, __artificial__))
 _mm512_store_epi64 (void *__P, __m512i __A)
 {
-  *(__m512i *) __P = __A;
+  *(__m512i_u *) __P = __A;
 }
 
 extern __inline void
@@ -520,14 +520,14 @@
 __attribute__ ((__gnu_inline__, __always_inline__, __artificial__))
 _mm512_store_si512 (void *__P, __m512i __A)
 {
-  *(__m512i *) __P = __A;
+  *(__m512i_u *) __P = __A;
 }
 
 extern __inline void
 __attribute__ ((__gnu_inline__, __always_inline__, __artificial__))
 _mm512_store_epi32 (void *__P, __m512i __A)
 {
-  *(__m512i *) __P = __A;
+  *(__m512i_u *) __P = __A;
 }
 
 extern __inline void
--- gcc-7.2.0/gcc/config/i386/avx512fintrin.h~	2017-10-22 15:32:30.000000000 +0000
+++ gcc-7.2.0/gcc/config/i386/avx512fintrin.h	2017-10-22 16:54:13.096698180 +0000
@@ -326,7 +326,7 @@
 __attribute__ ((__gnu_inline__, __always_inline__, __artificial__))
 _mm512_load_pd (void const *__P)
 {
-  return *(__m512d *) __P;
+  return *(__m512d_u *) __P;
 }
 
 extern __inline __m512d
@@ -367,7 +367,7 @@
 __attribute__ ((__gnu_inline__, __always_inline__, __artificial__))
 _mm512_load_ps (void const *__P)
 {
-  return *(__m512 *) __P;
+  return *(__m512_u *) __P;
 }
 
 extern __inline __m512
@@ -427,7 +427,7 @@
 __attribute__ ((__gnu_inline__, __always_inline__, __artificial__))
 _mm512_load_epi64 (void const *__P)
 {
-  return *(__m512i *) __P;
+  return *(__m512i_u *) __P;
 }
 
 extern __inline __m512i
@@ -487,14 +487,14 @@
 __attribute__ ((__gnu_inline__, __always_inline__, __artificial__))
 _mm512_load_si512 (void const *__P)
 {
-  return *(__m512i *) __P;
+  return *(__m512i_u *) __P;
 }
 
 extern __inline __m512i
 __attribute__ ((__gnu_inline__, __always_inline__, __artificial__))
 _mm512_load_epi32 (void const *__P)
 {
-  return *(__m512i *) __P;
+  return *(__m512i_u *) __P;
 }
 
 extern __inline __m512i
