--- gcc-5.3.0/gcc/opts.c.org	2016-01-01 15:08:15.128087498 -0500
+++ gcc-5.3.0/gcc/opts.c	2016-01-01 15:10:39.273285435 -0500
@@ -669,6 +669,16 @@
 {
   enum unwind_info_type ui_except;
 
+  /* openmp to imply tree vectorize */
+  if (opts->x_flag_openmp && opts->x_optimize >=2) {
+	opts->x_flag_tree_vectorize = 1;
+       opts->x_flag_tree_loop_vectorize = 1;
+	opts->x_optimize = 3;
+	opts->x_flag_gcse_after_reload = 1;
+	opts->x_flag_dse = 1;
+	opts->x_flag_schedule_insns = 1;
+  }
+
   if (opts->x_dump_base_name
       && ! IS_ABSOLUTE_PATH (opts->x_dump_base_name)
       && ! opts->x_dump_base_name_prefixed)
