--- gcc-5.3.0/gcc/opts.c~	2015-02-17 21:25:56.000000000 +0000
+++ gcc-5.3.0/gcc/opts.c	2016-04-11 18:14:36.569419793 +0000
@@ -569,7 +569,8 @@
 			       "integer, %<g%>, %<s%> or %<fast%>");
 	      else
 		{
-		  opts->x_optimize = optimize_val;
+                  if (optimize_val > opts->x_optimize)
+  		    opts->x_optimize = optimize_val;
 		  if ((unsigned int) opts->x_optimize > 255)
 		    opts->x_optimize = 255;
 		  opts->x_optimize_size = 0;
