--- gcc-7.3.0/gcc/config/i386/i386-c.c~	2017-02-17 18:11:28.000000000 +0000
+++ gcc-7.3.0/gcc/config/i386/i386-c.c	2018-02-18 17:02:34.008978140 +0000
@@ -358,31 +358,31 @@
     def_or_undef (parse_in, "__AVX__");
   if (isa_flag & OPTION_MASK_ISA_AVX2)
     def_or_undef (parse_in, "__AVX2__");
-  if (isa_flag & OPTION_MASK_ISA_AVX512F)
+  if (isa_flag & OPTION_MASK_ISA_AVX512F && !TARGET_PREFER_AVX256)
     def_or_undef (parse_in, "__AVX512F__");
-  if (isa_flag & OPTION_MASK_ISA_AVX512ER)
+  if (isa_flag & OPTION_MASK_ISA_AVX512ER && !TARGET_PREFER_AVX256)
     def_or_undef (parse_in, "__AVX512ER__");
-  if (isa_flag & OPTION_MASK_ISA_AVX512CD)
+  if (isa_flag & OPTION_MASK_ISA_AVX512CD && !TARGET_PREFER_AVX256)
     def_or_undef (parse_in, "__AVX512CD__");
-  if (isa_flag & OPTION_MASK_ISA_AVX512PF)
+  if (isa_flag & OPTION_MASK_ISA_AVX512PF && !TARGET_PREFER_AVX256)
     def_or_undef (parse_in, "__AVX512PF__");
-  if (isa_flag & OPTION_MASK_ISA_AVX512DQ)
+  if (isa_flag & OPTION_MASK_ISA_AVX512DQ && !TARGET_PREFER_AVX256)
     def_or_undef (parse_in, "__AVX512DQ__");
-  if (isa_flag & OPTION_MASK_ISA_AVX512BW)
+  if (isa_flag & OPTION_MASK_ISA_AVX512BW && !TARGET_PREFER_AVX256)
     def_or_undef (parse_in, "__AVX512BW__");
-  if (isa_flag & OPTION_MASK_ISA_AVX512VL)
+  if (isa_flag & OPTION_MASK_ISA_AVX512VL && !TARGET_PREFER_AVX256)
     def_or_undef (parse_in, "__AVX512VL__");
-  if (isa_flag & OPTION_MASK_ISA_AVX512VBMI)
+  if (isa_flag & OPTION_MASK_ISA_AVX512VBMI && !TARGET_PREFER_AVX256)
     def_or_undef (parse_in, "__AVX512VBMI__");
-  if (isa_flag & OPTION_MASK_ISA_AVX512IFMA)
+  if (isa_flag & OPTION_MASK_ISA_AVX512IFMA && !TARGET_PREFER_AVX256)
     def_or_undef (parse_in, "__AVX512IFMA__");
-  if (isa_flag2 & OPTION_MASK_ISA_AVX5124VNNIW)
+  if (isa_flag2 & OPTION_MASK_ISA_AVX5124VNNIW && !TARGET_PREFER_AVX256)
     def_or_undef (parse_in, "__AVX5124VNNIW__");
   if (isa_flag2 & OPTION_MASK_ISA_SGX)
     def_or_undef (parse_in, "__SGX__");
-  if (isa_flag2 & OPTION_MASK_ISA_AVX5124FMAPS)
+  if (isa_flag2 & OPTION_MASK_ISA_AVX5124FMAPS && !TARGET_PREFER_AVX256)
     def_or_undef (parse_in, "__AVX5124FMAPS__");
-  if (isa_flag2 & OPTION_MASK_ISA_AVX512VPOPCNTDQ)
+  if (isa_flag2 & OPTION_MASK_ISA_AVX512VPOPCNTDQ && !TARGET_PREFER_AVX256)
     def_or_undef (parse_in, "__AVX512VPOPCNTDQ__");
   if (isa_flag & OPTION_MASK_ISA_FMA)
     def_or_undef (parse_in, "__FMA__");
