commit bc1421a83844007df2e17c592721a6c1bc772179
Author: H.J. Lu <hjl.tools@gmail.com>
Date:   Tue Feb 6 13:31:15 2018 -0800

    Add -mzero-caller-saved-regs

diff --git a/gcc/config/i386/i386.c b/gcc/config/i386/i386.c
index eca3ea68fb3..5ef5b3c71f5 100644
--- a/gcc/config/i386/i386.c
+++ b/gcc/config/i386/i386.c
@@ -7155,6 +7155,10 @@ ix86_set_func_type (tree fndecl)
 	  if (lookup_attribute ("no_caller_saved_registers",
 				TYPE_ATTRIBUTES (TREE_TYPE (fndecl))))
 	    cfun->machine->no_caller_saved_registers = true;
+	  if (ix86_zero_caller_saved_regs
+	      || lookup_attribute ("zero_caller_saved_regs",
+				   TYPE_ATTRIBUTES (TREE_TYPE (fndecl))))
+	    cfun->machine->zero_caller_saved_regs = true;
 	}
     }
 }
@@ -14540,7 +14544,7 @@ ix86_expand_prologue (void)
       insn = emit_insn (gen_set_got (pic));
       RTX_FRAME_RELATED_P (insn) = 1;
       add_reg_note (insn, REG_CFA_FLUSH_QUEUE, NULL_RTX);
-      emit_insn (gen_prologue_use (pic));
+      emit_insn (gen_pro_epilogue_use (pic));
       /* Deleting already emmitted SET_GOT if exist and allocated to
 	 REAL_PIC_OFFSET_TABLE_REGNUM.  */
       ix86_elim_entry_set_got (pic);
@@ -14569,7 +14573,7 @@ ix86_expand_prologue (void)
      Further, prevent alloca modifications to the stack pointer from being
      combined with prologue modifications.  */
   if (TARGET_SEH)
-    emit_insn (gen_prologue_use (stack_pointer_rtx));
+    emit_insn (gen_pro_epilogue_use (stack_pointer_rtx));
 }
 
 /* Emit code to restore REG using a POP insn.  */
@@ -14642,6 +14646,84 @@ ix86_emit_restore_regs_using_pop (void)
       ix86_emit_restore_reg_using_pop (gen_rtx_REG (word_mode, regno));
 }
 
+/* Emit code to zero touched caller-saved general registers upon
+   function return.  */
+
+static void
+ix86_emit_zero_caller_saved_regs (void)
+{
+  /* No need to zero caller-saved registers in main ().  Don't zero
+     caller-saved registers if __builtin_eh_return is called since it
+     isn't a normal function return.  */
+  if (!cfun->machine->zero_caller_saved_regs
+      || crtl->calls_eh_return
+      || MAIN_NAME_P (DECL_NAME (current_function_decl)))
+    return;
+
+  edge e;
+  edge_iterator ei;
+
+  unsigned int regno;
+
+  HARD_REG_SET live_outgoing_regs;
+  CLEAR_HARD_REG_SET (live_outgoing_regs);
+
+  /* Check for live outgoing registers.  */
+  FOR_EACH_EDGE (e, ei, EXIT_BLOCK_PTR_FOR_FN (cfun)->preds)
+    {
+      basic_block bb = e->src;
+      bitmap live_out = df_get_live_out (bb);
+
+      for (regno = 0; regno < FIRST_PSEUDO_REGISTER; regno++)
+	{
+	  /* Only zero general registers.  */
+	  if (!GENERAL_REGNO_P (regno))
+	    continue;
+
+	  /* No need to check it again if it is live.  */
+	  if (TEST_HARD_REG_BIT (live_outgoing_regs, regno))
+	    continue;
+
+	  /* A register is considered LIVE if
+	     1. It is a fixed register.
+	     2. If isn't a caller-saved register.
+	     3. If it is a live outgoing register.
+	     4. If it is neverd used.
+	     */
+	  if (fixed_regs[regno]
+	      || !call_used_regs[regno]
+	      || REGNO_REG_SET_P (live_out, regno)
+	      || !df_regs_ever_live_p (regno))
+	    SET_HARD_REG_BIT (live_outgoing_regs, regno);
+	}
+    }
+
+  rtx zero = NULL_RTX;
+
+  for (regno = 0; regno < FIRST_PSEUDO_REGISTER; regno++)
+    {
+      if (!GENERAL_REGNO_P (regno))
+	continue;
+
+      if (TEST_HARD_REG_BIT (live_outgoing_regs, regno))
+	continue;
+
+      /* Zero out dead caller-saved register.  We only need to zero the
+	 lower 32 bits.  */
+      rtx reg = gen_rtx_REG (SImode, regno);
+      if (zero == NULL_RTX)
+	{
+	  zero = reg;
+	  emit_move_insn (reg, const0_rtx);
+	}
+      else {
+	emit_move_insn (reg, zero); zero = NULL_RTX;
+      }
+      /* Mark it in use  */
+      emit_insn (gen_pro_epilogue_use (reg));
+    }
+}
+
 /* Emit code and notes for the LEAVE instruction.  */
 
 static void
@@ -15089,10 +15171,16 @@ ix86_expand_epilogue (int style)
 	  emit_jump_insn (gen_simple_return_indirect_internal (ecx));
 	}
       else
-	emit_jump_insn (gen_simple_return_pop_internal (popc));
+	{
+	  ix86_emit_zero_caller_saved_regs ();
+	  emit_jump_insn (gen_simple_return_pop_internal (popc));
+	}
     }
   else
-    emit_jump_insn (gen_simple_return_internal ());
+    {
+      ix86_emit_zero_caller_saved_regs ();
+      emit_jump_insn (gen_simple_return_internal ());
+    }
 
   /* Restore the state back to the state from the prologue,
      so that it's correct for the next epilogue.  */
@@ -46576,6 +46664,8 @@ static const struct attribute_spec ix86_attribute_table[] =
     ix86_handle_fndecl_attribute, false },
   { "function_return", 1, 1, true, false, false,
     ix86_handle_fndecl_attribute, false },
+  { "zero_caller_saved_regs", 0, 0, true, false, false,
+    NULL, false },
 
   /* End element.  */
   { NULL,        0, 0, false, false, false, NULL, false }
diff --git a/gcc/config/i386/i386.h b/gcc/config/i386/i386.h
index e7da790f1b1..42a0dcae374 100644
--- a/gcc/config/i386/i386.h
+++ b/gcc/config/i386/i386.h
@@ -2623,6 +2623,10 @@ struct GTY(()) machine_function {
      the "interrupt" or "no_caller_saved_registers" attribute.  */
   BOOL_BITFIELD no_caller_saved_registers : 1;
 
+  /* If true, the current function should zero out touched caller-saved
+     general registers upon function return.  */
+  BOOL_BITFIELD zero_caller_saved_regs : 1;
+
   /* If true, there is register available for argument passing.  This
      is used only in ix86_function_ok_for_sibcall by 32-bit to determine
      if there is scratch register available for indirect sibcall.  In
diff --git a/gcc/config/i386/i386.md b/gcc/config/i386/i386.md
index 93ebbbd87bc..04304f4dd9f 100644
--- a/gcc/config/i386/i386.md
+++ b/gcc/config/i386/i386.md
@@ -206,7 +206,7 @@
   UNSPECV_STACK_PROBE
   UNSPECV_PROBE_STACK_RANGE
   UNSPECV_ALIGN
-  UNSPECV_PROLOGUE_USE
+  UNSPECV_PRO_EPILOGUE_USE
   UNSPECV_SPLIT_STACK_RETURN
   UNSPECV_CLD
   UNSPECV_NOPS
@@ -12280,8 +12280,8 @@
 
 ;; As USE insns aren't meaningful after reload, this is used instead
 ;; to prevent deleting instructions setting registers for PIC code
-(define_insn "prologue_use"
-  [(unspec_volatile [(match_operand 0)] UNSPECV_PROLOGUE_USE)]
+(define_insn "pro_epilogue_use"
+  [(unspec_volatile [(match_operand 0)] UNSPECV_PRO_EPILOGUE_USE)]
   ""
   ""
   [(set_attr "length" "0")])
diff --git a/gcc/config/i386/i386.opt b/gcc/config/i386/i386.opt
index b90da9f89ec..6dbd96c9db5 100644
--- a/gcc/config/i386/i386.opt
+++ b/gcc/config/i386/i386.opt
@@ -955,3 +955,7 @@ Enum(indirect_branch) String(thunk-extern) Value(indirect_branch_thunk_extern)
 mindirect-branch-register
 Target Report Var(ix86_indirect_branch_register) Init(0)
 Force indirect call and jump via register.
+
+mzero-caller-saved-regs
+Target Report Var(ix86_zero_caller_saved_regs) Init(0)
+Zero out touched caller-saved general registers upon function return.
